README.md
# HdfsCLI [![CI](https://github.com/mtth/hdfs/actions/workflows/ci.yml/badge.svg)](https://github.com/mtth/hdfs/actions/workflows/ci.yml) [![Pypi badge](https://badge.fury.io/py/hdfs.svg)](https://pypi.python.org/pypi/hdfs/) [![Downloads badge](https://img.shields.io/pypi/dm/hdfs.svg)](https://pypistats.org/packages/hdfs)

API and command line interface for HDFS.

```
$ hdfscli --alias=dev

Welcome to the interactive HDFS python shell.
The HDFS client is available as `CLIENT`.

In [1]: CLIENT.list('models/')
Out[1]: ['1.json', '2.json']

In [2]: CLIENT.status('models/2.json')
Out[2]: {
  'accessTime': 1439743128690,
  'blockSize': 134217728,
  'childrenNum': 0,
  'fileId': 16389,
  'group': 'supergroup',
  'length': 48,
  'modificationTime': 1439743129392,
  'owner': 'drwho',
  'pathSuffix': '',
  'permission': '755',
  'replication': 1,
  'storagePolicy': 0,
  'type': 'FILE'
}

In [3]: with CLIENT.read('models/2.json', encoding='utf-8') as reader:
  ...:     from json import load
  ...:     model = load(reader)
  ...:
```

## Features

* Python 3 bindings for the [WebHDFS][] (and [HttpFS][]) API,
  supporting both secure and insecure clusters.
* Command line interface to transfer files and start an interactive client
  shell, with aliases for convenient namenode URL caching.
* Additional functionality through optional extensions:

  + `avro`, to [read and write Avro files directly from HDFS][].
  + `dataframe`, to [load and save Pandas dataframes][].
  + `kerberos`, to [support Kerberos authenticated clusters][].

See the [documentation][] to learn more.

## Getting started

```sh
$ pip install hdfs
```

Then hop on over to the [quickstart][] guide. A [Conda
feedstock](https://github.com/conda-forge/python-hdfs-feedstock) is also
available.

## Testing

HdfsCLI is tested against both [WebHDFS][] and [HttpFS][]. There are two ways
of running tests (see `scripts/` for helpers to set up a test HDFS cluster):

```sh
$ HDFSCLI_TEST_URL=http://localhost:50070 pytest # Using a namenode's URL.
$ HDFSCLI_TEST_ALIAS=dev pytest # Using an alias.
```

## Contributing

We'd love to hear what you think on the [issues][] page. Pull requests are also
most welcome!

[HttpFS]: http://hadoop.apache.org/docs/current/hadoop-hdfs-httpfs/
[WebHDFS]: http://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/WebHDFS.html
[read and write Avro files directly from HDFS]: https://hdfscli.readthedocs.io/en/latest/api.html#module-hdfs.ext.avro
[load and save Pandas dataframes]: https://hdfscli.readthedocs.io/en/latest/api.html#module-hdfs.ext.dataframe
[support Kerberos authenticated clusters]: https://hdfscli.readthedocs.io/en/latest/api.html#module-hdfs.ext.kerberos
[documentation]: https://hdfscli.readthedocs.io/
[quickstart]: https://hdfscli.readthedocs.io/en/latest/quickstart.html
[issues]: https://github.com/mtth/hdfs/issues


CHANGES
HdfsCLI
=======

Version 2.0 (2015/20/08)
------------------------

* Add python 3 support (the Kerberos extension's requirements must however be 
  manually installed).
* Allow specifying relative client roots. These will be assumed relative to the 
  user's home directory.
* Add several client methods: `makedirs`, `set_times`, `checksum`, 
  `set_replication`, etc.
* Add `progress` argument to `Client.read`, `Client.upload`, and 
  `Client.download`. Also add a `chunk_size` argument to the latter to allow 
  better tracking.
* Add `strict` option to `Client.status` and `Client.content` to perform path 
  existence checks.
* `Client.write` now can be used as a context manager (returning a file-like 
  object).
* `Client.read` and `Client.write` can now return file-like objects (supporting 
  `read` and `write` calls respectively).
* Improve robustness of `KerberosClient`. In particular add the 
  `max_concurrency` parameter which can be tuned to prevent authentication 
  errors when too many simultaneous requests are being made. Along with the new 
  delay parameter, this lets us remove the timeouts in `Client.download` and 
  `Client.upload`, which both simplify and speed up these functions.
* Add `Config` class which handles all CLI configuration (e.g. aliases and 
  logging).
* Add `autoload.modules` and `autoload.paths` configuration options.
* Rename alias configuration sections to `ALIAS.alias` (the old format, 
  `ALIAS_alias` is still supported).
* Add `--verbose` option to CLI, enabling logging at various levels.
* Switch Avro extension to using `fastavro`. This speeds up `AvroWriter` and 
  `AvroReader` by a significant amount (~5 times faster on a reasonable 
  connection).
* Add `write` command to Avro CLI.
* Remove CSV support for dataframe extension.

Breaking changes:

* Change default configuration file path (to `~/.hdfscli.cfg`).
* Change location of `default.alias` option in configuration file (from command 
  specific section to `global`).
* Add `session` argument to `Client` and remove newly redundant parameters 
  (e.g. `verify`, `cert`).
* Remove `Client.from_alias` method (delegated to the new `Config` class). The 
  `Client.from_options` method is now public (renamed from 
  `Client._from_options`).
* Change default entry point name to `hdfscli` to avoid clashing with Hadoop 
  HDFS script.
* `Client.delete` now returns a boolean indicating success (rather than failing 
  if the path was already empty).
* `Client.read` must now be used inside a `with` block. This ensures that 
  connections are properly closed. The context manager now returns a file like 
  object (useful for composing with other functions, e.g. to decode Avro). 
  Setting `chunk_size` to a positive value will make it return a generator 
  similar to the previous behavior.
* Rename `Client.set_permissions` to `Client.set_permission` to make 
  `permission` argument uniform across `Client` methods (always singular, 
  consistent with WebHDFS API).
* `Client.parts` will now throw an error if called on a normal file.
* Make most client attributes private (e.g. `cert`, `timeout`, etc.), except 
  `url` and `root`.
* Remove `--` prefix from CLI commands. Also simplify CLI to only interactive, 
  download and upload commands (write and read behavior can be achieved by 
  passing '-' as local path). The `Client` API changes should make it more 
  convenient to perform these from a python shell.
* Rename several CLI options (e.g. `--log`, `--force`, `--version`).
* Change meaning of `n_threads` option in `Client.download` and 
  `Client.upload`. `0` now means one thread per part-file rather than a single 
  thread.
* Change `Client.walk` to be consistent with `os.walk`. Also change meaning of
  `depth` option (`0` being unlimited).
* Add `status` option to `Client.list`, `Client.walk`, and `Client.parts`. By
  default these functions now only return the names of the relevant files and
  folders.
* Remove `Client.append` method (replaced by `append` keyword argument to 
  `Client.write`).
* Symbols exported by extensions aren't imported in the main `hdfs` module 
  anymore. This removes the need for some custom error handling (when 
  dependency requirements weren't met).
* Remove Bash autocompletion file (for now).
* Remove compatibility layer for entry point configuration (i.e. 
  `HDFS_ENTRY_POINT` isn't supported anymore).


Version 1.4.0 (2015/07/24)
--------------------------

* Add support for download and upload of arbitrary folders.
* Deprecate `Client.append` (in favor of `append` argument to `Client.write`).


Version 1.1.0 (2015/06/23)
--------------------------

* Rename Avro extension entry point to `hdfs-avro`.


Version 1.0.1 (2015/06/17)
--------------------------

* Added support for Windows.
* Added support for remote filepaths with `=` characters.


Version 0.3.0 (2014/11/14)
--------------------------

* Added `--interactive` command.

Breaking changes:

* Renamed `--info` command to `--list`.
* Made `--interactive` the new default command.


Version 0.2.6 (2014/08/04)
--------------------------

* Added parallelized downloading.
* Added Avro-format reading and writing.
* Added `hdfs.ext.dataframe` extension.


Version 0.2.0 (2014/04/26)
--------------------------

* Added `Client.status` and `Client.content` methods.
* Added callback to `Client.write`.

Breaking changes:

* Removed content from `Client.walk`.
* Simplified CLI. All download and uploads are normalized through standard in, 
  and standard out.


Version 0.1.0 (2014/03/25)
--------------------------

* Initial release


hdfs/__main__.py
#!/usr/bin/env python
# encoding: utf-8

"""HdfsCLI: a command line interface for HDFS.

Usage:
  hdfscli [interactive] [-a ALIAS] [-v...]
  hdfscli download [-fsa ALIAS] [-v...] [-t THREADS] HDFS_PATH LOCAL_PATH
  hdfscli upload [-sa ALIAS] [-v...] [-A | -f] [-r REPLICATION] [-t THREADS] LOCAL_PATH HDFS_PATH
  hdfscli -L | -V | -h

Commands:
  download                      Download a file or folder from HDFS. If a
                                single file is downloaded, - can be
                                specified as LOCAL_PATH to stream it to
                                standard out.
  interactive                   Start the client and expose it via the python
                                interpreter (using iPython if available).
  upload                        Upload a file or folder to HDFS. - can be
                                specified as LOCAL_PATH to read from standard
                                in.

Arguments:
  HDFS_PATH                     Remote HDFS path.
  LOCAL_PATH                    Path to local file or directory.

Options:
  -A --append                   Append data to an existing file. Only supported
                                if uploading a single file or from standard in.
  -L --log                      Show path to current log file and exit.
  -V --version                  Show version and exit.
  -a ALIAS --alias=ALIAS        Alias of namenode to connect to.
  -f --force                    Allow overwriting any existing files.
  -s --silent                   Don't display progress status.
  -t THREADS --threads=THREADS  Number of threads to use for parallelization.
                                0 allocates a thread per file. [default: 0]
  -v --verbose                  Enable log output. Can be specified up to three
                                times (increasing verbosity each time).
  -r REPLICATION --replication=REPLICATION  Set the expected HDFS replication count [default: 3]

Examples:
  hdfscli -a prod /user/foo
  hdfscli download features.avro dat/
  hdfscli download logs/1987-03-23 - >>logs
  hdfscli upload -f - data/weights.tsv <weights.tsv
  hdfscli upload -f -r 1 - data/weights.tsv <weights.tsv

HdfsCLI exits with return status 1 if an error occurred and 0 otherwise.

"""

from . import __version__
from .config import Config, NullHandler, catch
from .util import HdfsError
from docopt import docopt
from threading import Lock
import logging as lg
import os
import os.path as osp
import sys


def parse_arg(args, name, parser, separator=None):
  """Parse command line argument, raising an appropriate error on failure.

  :param args: Arguments dictionary.
  :param name: Name of option to look up.
  :param parser: Function to parse option.
  :param separator: For parsing lists.

  """
  value = args[name]
  if not value:
    return
  try:
    if separator and separator in value:
      return [parser(part) for part in value.split(separator) if part]
    else:
      return parser(value)
  except ValueError:
    raise HdfsError('Invalid %r option: %r.', name, args[name])

def configure_client(command, args, config=None):
  """Instantiate configuration from arguments dictionary.

  :param command: Command name, used to set up the appropriate log handler.
  :param args: Arguments returned by `docopt`.
  :param config: CLI configuration, used for testing.

  If the `--log` argument is set, this method will print active file handler
  paths and exit the process.

  """
  logger = lg.getLogger()
  logger.setLevel(lg.DEBUG)
  lg.getLogger('requests_kerberos.kerberos_').setLevel(lg.INFO)
  # TODO: Filter only at handler level.
  if not config:
    levels = {0: lg.ERROR, 1: lg.WARNING, 2: lg.INFO}
    config = Config(stream_log_level=levels.get(args['--verbose'], lg.DEBUG))
  handler = config.get_log_handler(command)
  if args['--log']:
    if isinstance(handler, NullHandler):
      sys.stdout.write('No log file active.\n')
      sys.exit(1)
    else:
      sys.stdout.write('%s\n' % (handler.baseFilename, ))
      sys.exit(0)
  logger.addHandler(handler)
  return config.get_client(args['--alias'])


class _Progress(object):

  """Progress tracker callback.

  :param nbytes: Total number of bytes that will be transferred.
  :param nfiles: Total number of files that will be transferred.
  :param writer: Writable file-object where the progress will be written.
    Defaults to standard error.

  """

  def __init__(self, nbytes, nfiles, writer=None):
    self._total_bytes = nbytes
    self._pending_files = nfiles
    self._writer = writer or sys.stderr
    self._downloading_files = 0
    self._complete_files = 0
    self._lock = Lock()
    self._data = {}

  def __call__(self, hdfs_path, nbytes):
    # TODO: Improve lock granularity.
    with self._lock:
      data = self._data
      if hdfs_path not in data:
        self._pending_files -= 1
        self._downloading_files += 1
      if nbytes == -1:
        self._downloading_files -= 1
        self._complete_files += 1
      else:
        data[hdfs_path] = nbytes
      if self._pending_files + self._downloading_files > 0:
        self._writer.write(
          '%3.1f%%\t[ pending: %d | downloading: %d | complete: %d ]   \r' %
          (
            100. * sum(data.values()) / self._total_bytes,
            self._pending_files,
            self._downloading_files,
            self._complete_files,
          )
        )
      else:
        self._writer.write('%79s\r' % ('', ))

  @classmethod
  def from_hdfs_path(cls, client, hdfs_path, writer=None):
    """Instantiate from remote path.

    :param client: HDFS client.
    :param hdfs_path: HDFS path.

    """
    content = client.content(hdfs_path)
    return cls(content['length'], content['fileCount'], writer=writer)

  @classmethod
  def from_local_path(cls, local_path, writer=None):
    """Instantiate from a local path.

    :param local_path: Local path.

    """
    if osp.isdir(local_path):
      nbytes = 0
      nfiles = 0
      for dpath, _, fnames in os.walk(local_path):
        for fname in fnames:
          nbytes += osp.getsize(osp.join(dpath, fname))
          nfiles += 1
    elif osp.exists(local_path):
      nbytes = osp.getsize(local_path)
      nfiles = 1
    else:
      raise HdfsError('No file found at: %s', local_path)
    return cls(nbytes, nfiles, writer=writer)

@catch(HdfsError)
def main(argv=None, client=None):
  """Entry point.

  :param argv: Arguments list.
  :param client: For testing.

  """
  args = docopt(__doc__, argv=argv, version=__version__)
  if not client:
    client = configure_client('hdfscli', args)
  elif args['--log']:
    raise HdfsError('Logging is only available when no client is specified.')
  hdfs_path = args['HDFS_PATH']
  local_path = args['LOCAL_PATH']
  n_threads = parse_arg(args, '--threads', int)
  replication_n = parse_arg(args, '--replication', int)
  force = args['--force']
  silent = args['--silent']
  if args['download']:
    chunk_size = 2 ** 16
    if local_path == '-':
      if not sys.stdout.isatty() and sys.stderr.isatty() and not silent:
        progress = _Progress.from_hdfs_path(client, hdfs_path)
      else:
        progress = None
      with client.read(
        hdfs_path,
        chunk_size=chunk_size,
        progress=progress,
      ) as reader:
        # https://stackoverflow.com/a/23932488/1062617
        stdout = getattr(sys.stdout, 'buffer', sys.stdout)
        for chunk in reader:
          stdout.write(chunk)
    else:
      if sys.stderr.isatty() and not silent:
        progress = _Progress.from_hdfs_path(client, hdfs_path)
      else:
        progress = None
      client.download(
        hdfs_path,
        local_path,
        overwrite=force,
        n_threads=n_threads,
        chunk_size=chunk_size,
        progress=progress,
      )
  elif args['upload']:
    append = args['--append']
    if local_path == '-':
      client.write(
        hdfs_path,
        (line for line in sys.stdin), # Doesn't work with stdin.
        append=append,
        overwrite=force,
        replication=replication_n,
      )
    else:
      if append:
        # TODO: Add progress tracking here.
        if osp.isfile(local_path):
          with open(local_path) as reader:
            client.write(hdfs_path, reader, append=True)
        else:
          raise HdfsError('Can only append when uploading a single file.')
      else:
        if sys.stderr.isatty() and not silent:
          progress = _Progress.from_local_path(local_path)
        else:
          progress = None
        client.upload(
          hdfs_path,
          local_path,
          overwrite=force,
          n_threads=n_threads,
          progress=progress,
          replication=replication_n,
        )
  else:
    banner = (
      '\n'
      'Welcome to the interactive HDFS python shell.\n'
      'The HDFS client is available as `CLIENT`.\n'
    )
    namespace = {'CLIENT': client}
    try:
      from IPython import embed
    except ImportError:
      from code import interact
      interact(banner=banner, local=namespace)
    else:
      embed(banner1=banner, user_ns=namespace)

if __name__ == '__main__':
  main()
